### DISCUSSION TITLE
Oauth2 on SPAs

### DISCUSSION THREAD

**Discussion by jimkoul** [fastapi/fastapi#5119] [Questions]
I'm wandering whats the best way to implement  login flows on a Single Page Application (React).
The best oauth2 flow seems to be "Authorization code with PKCE".
Does fastapi support that flow?

Also,a bit off topic maybe,whats the best way to store the tokens on SPAs or native apps?
Is it localStorage secure or not?especially for refresh tokens

**Comment by EssaAlshammri** [comment]
check this out
https://codeburst.io/localstorage-vs-cookies-all-you-need-to-know-about-storing-jwt-tokens-securely-in-the-front-end-70dc0a9b3ad3

**Reply by jimkoul** [reply]
i have already read this article too,to be honest.Very good article.

Though,im still not sure how to implement this in real life.

I build a demo fastapi project using built-in oauth feature but at the end of the day the access token or session must be stored somewhere on the clients in order to work as expected.

An other solution i read is the Backend For Frontend (BFF) that, if im correct, is something like a proxy backend code.
The proxy make all the oauth flows as a result of the Client command.The access token and refresh token are stored there and the BFF returns a session http only,secure cookie to the client.The client use this on every request instead of the Authorisation Bearer Header.
All requests are passed and proxied by the BFF transating the cookie to access token before hit the actual backend api.

What about that solution?

Do anyone have a real-life production app/website using this  or another secure flow?
